src/els_compiler_diagnostics.erl

```diff
@@ -71,6 +71,8 @@ compile(Uri) ->
   Includes = [ {i, IncludePath}
                || IncludePath <- els_config:get(include_paths)
              ],
+  %% Extra paths to load beam files from, for e.g. behaviours
+  [code:add_patha(Dir) || Dir <- els_config:get(pa_dirs)],
```

Comment:
The approach proposed relies on loading binaries onto the server which have been produced by some external tool. This could be problematic and lead to a series of edge cases which could be confusing for the user (e.g. if different compiler options have been used).

Imagine the following scenario: the developer works on a behaviour and its implementation at the same time (two buffers open). When saving the behaviour, changes are not reflected on the server. The user should manually re-trigger compilation using an external tool (rebar3 in this case) for the server to pick them up.

I'd suggest an alternative approach. If the current module contains a behaviour definition (and, eventually, a parse transform), we look for the behaviour URI. We compile the behaviour and load the binary into memory (no need to persist the .beam to disk). The compilation of the current module would then succeed. We can revert the load if we want to avoid persisting the side effect.

This whole thing also leads to a reflection. As soon as we start operating on binary files (persisted or not), we need to/should ensure that the language server node is running the version of OTP specified by the `erlang_ls.config` or issues may arise. This is not the case at the moment, where the OTP version specified by the user is only used for indexing, not to select the runtime. Dynamically selecting the run-time could also problematic, since the server would have been compiled with one version of OTP, but use a different one at runtime. This was one of the major pain points with the EDTS project.

Reply:
Something like:

```
{ok, my_module, B} = compile:file("/path/to/my_module.erl", [binary]).
code:load_binary(my_module, no_name, B).
```

Reply:
An alternative would be to filter out the `undefined_behaviour` warnings and re-implement the verification logic based on just parsing the files inside of erlang_ls without actually loading any modules

Reply:
I presume the approach should be to do the compile as now, then pick up "undefined_behaviour" warnings and use that to trigger an attempted load of the behaviour and recompile.  Being careful to avoid infinite loops.

Reply:
> When saving the behaviour, changes are not reflected on the server.

Thinking about this, we currently invoke the compiler for the file to generate diagnostics. Are you proposing we *never* compile to beam, to avoid messing with the underlying build system. Which makes sense to me.  Or even do a cheap compile most times and avoid code gen.

Reply:
Which I see already happens, via `basic_validation` flag.